tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Azure/Functions/Isolated/GrpcMessageConversionExtensionsToRpcHttpIntegration.cs

```diff
@@ -31,17 +33,62 @@ namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Azure.Functions;
+        var capabilities = state.State.DuckCast<IGrpcCapabilities>();
+        if (capabilities is null || returnValue is null)
+        {
+            // Something went wrong, this shouldn't happen
+            return returnValue;
+        }
+
+        var tracer = Tracer.Instance;
+        if (!tracer.Settings.IsIntegrationEnabled(AzureFunctionsCommon.IntegrationId)
+         || tracer.ActiveScope is not Scope { Span: { OperationName: AzureFunctionsCommon.OperationName } span })
+        {
+            return returnValue;
+        }
+
+        // The HTTP request represented by TypedData is essentially a duplicate of the original incoming
+        // request that was received by func.exe. This is used to create a span representing
+        // the request from the client. The typed data is then sent by the GRPC connection
+        // to the functions app and is used to invoke the actual function. We intercept that
+        // in the functions app and use it to create a span representing the actual work of the app.
+        // In order for the span hierarchy/parenting to work correctly, we need to replace the parentID
+        // in the GRPC http request representation, which is what we're doing here by overwriting all
+        // the existing datadog headers
+        //
+        // However, when using the AspNetCore integration, things work a bit differently. The
+        // func.exe app instead primarily _proxies_ the HTTP request (if it is an HTTP trigger) to the functions
+        // app, instead of sending the bulk of the context as a gRPC message. This means that we _shouldn't_ inject the
+        // context into the gRPC request, because the context is already present in the incoming HTTP request, and is
+        // used preferentially.
+        //
+        // What's more, in the case of HTTP triggers with proxying enabled, the TypedData returnValue returned from
+        // this is method is a shared object, so mutating it can cause issues with other parts of the system.
+        // See https://github.com/Azure/azure-functions-host/blob/420a4686802612857cae35cefea2b685283507c9/src/WebJobs.Script.Grpc/MessageExtensions/GrpcMessageConversionExtensions.cs#L104-L126
+
+        var isHttpProxying = !string.IsNullOrEmpty(capabilities.GetCapabilityState("HttpUri"));
+        var requiresRouteParameters = !string.IsNullOrEmpty(capabilities.GetCapabilityState("RequiresRouteParameters"));
+        var useNullableHeaders = !string.IsNullOrEmpty(capabilities.GetCapabilityState("UseNullableValueDictionaryForHttp"));
+
+        // When proxying, this method returns a singleton value that we must not update, so we create a new one
+        // If we're not proxying, we can safely inject the context into the provided gRPC request
+        var typedData = isHttpProxying && !requiresRouteParameters
+                            ? TypedDataHelper<TReturn>.CreateTypedData()
+                            : returnValue.DuckCast<ITypedData>();
```

Comment:
I wonder if there's a way to identify if returnValue is the singleton value, to have more robust code? (so that we don't have to rely on conditions such as `isHttpProxying` that may evolve in the future).

Sounds unlikely but I had to ask.

Reply:
Discussed offline. We _could_ check this, but duck typing the extension instance, accessing the `Empty` field, and comparing it to `returnValue`. The difficulty is that doing this isn't necessary any less brittle - if they move/rename the `Empty` field then we might _not_ detect it's a "null" entry. In that case we would create a new instance of the typedData, stomping any other potential changes made to the incoming object.

So in conclusion
- This sucks
- The alternative sucks
- Doing both sucks

😅 So we just need to be sure to keep on top of any updates, and be particularly wary of changes in this method or to how capabilities are calculated